gpjax/decision_making/decision_maker.py

```diff
@@ -0,0 +1,254 @@
+    posterior_handlers: Dict[str, PosteriorHandler]
+    datasets: Dict[str, Dataset]
+    acquisition_function_builder: AbstractAcquisitionFunctionBuilder
+    acquisition_maximizer: AbstractAcquisitionMaximizer
+    key: KeyArray
+    post_ask: List[Callable] = field(
+        default_factory=list
+    )  # Specific type is List[Callable[[DecisionMaker, Float[Array, ["1 D"]]], None]] but causes Beartype issues
+    post_tell: List[Callable] = field(
+        default_factory=list
+    )  # Specific type is List[Callable[[DecisionMaker], None]] but causes Beartype issues
+
+    @abstractmethod
+    def ask(self, key: KeyArray) -> Float[Array, "1 D"]:
+        """
+        In a typical decision making setup this will use the
+        `acquisition_function_builder` to form an acquisition function and then return
+        the point which maximizes the acquisition function using the
+        `acquisition_maximizer` as the point to be queried next.
+
+        Args:
+            key (KeyArray): JAX PRNG key for controlling random state.
+
+        Returns:
+            Float[Array, "1 D"]: Point to be queried next
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def tell(self, observation_datasets: Mapping[str, Dataset], key: KeyArray):
+        """
+        Tell decision maker about new observations. In a typical decision making setup
+        we will update the datasets and posteriors with the new observations.
+
+        Args:
+            observation_datasets (Mapping[str, Dataset]): Dictionary of datasets
+                containing new observations. Tags are used to distinguish datasets, and
+                correspond to tags in `posterior_handlers` in a typical setup.
+            key (KeyArray): JAX PRNG key for controlling random state.
+        """
+        raise NotImplementedError
+
+
+@dataclass
+class DecisionMaker(AbstractDecisionMaker):
+    """
+    DecisionMaker class which handles the core decision making loop in a typical setup. The
+    decision making loop is split into two key steps, `ask` and `tell`. The `ask`
+    step forms an `AcquisitionFunction` from the current `posteriors` and `datasets` and
+    returns the point which maximises it. It also stores the formed acquisition function
+    under the attribute `self.current_acquisition_function` so that it can be called,
+    for instance for plotting, after the `ask` function has been called. The `tell` step
+    adds a newly queried point to the `datasets` and updates the `posteriors`.
+
+    This can be run as a typical ask-tell loop, or the `run` method can be used to run
+    the decision making loop for a fixed number of steps. Moreover, the `run` method executes
+    the functions in `post_ask` and `post_tell` after each ask and tell step
+    respectively. This enables the user to add custom functionality, such as the ability
+    to plot values of interest during the optimization process.
+    """
+
+    def __post_init__(self):
```

Comment:
Hadn't included originally in case people wanted a decision maker which didn’t optimise all the models/there was some case where people didn’t need to have exactly the same datasets as models. But I suppose in this case they can just override `__post_init__`, so will move to the abstract class.